Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/SearchTry.hs

```diff
@@ -118,32 +120,70 @@ initiateDriverSearchBatch searchBatchInput@DriverSearchBatchInput {..} = do
   searchTry <- createNewSearchTry
   driverPoolConfig <- getDriverPoolConfig searchReq.merchantOperatingCityId searchTry.vehicleServiceTier searchTry.tripCategory (fromMaybe SL.Default searchReq.area) searchReq.estimatedDistance (Just (TransactionId (Id searchReq.transactionId))) searchReq
   goHomeCfg <- CGHC.findByMerchantOpCityId searchReq.merchantOperatingCityId (Just (TransactionId (Id searchReq.transactionId)))
-  singleBatchProcessingTempDelay <- asks (.singleBatchProcessingTempDelay)
   now <- getCurrentTime
+  singleBatchProcessingTempDelay <- asks (.singleBatchProcessingTempDelay)
   let scheduleTryTimes = secondsToNominalDiffTime . Seconds <$> driverPoolConfig.scheduleTryTimes
       instantReallocation = maybe True (\scheduleTryTime -> diffUTCTime searchReq.startTime now <= scheduleTryTime) (listToMaybe scheduleTryTimes)
   if not searchTry.isScheduled || (instantReallocation && isRepeatSearch)
     then do
+      let remainingTime = diffUTCTime searchReq.startTime now
+          lastScheduledTime = fromMaybe 1800 (findPrevScheduleTime remainingTime scheduleTryTimes)
+          buffer = intToNominalDiffTime (driverPoolConfig.maxNumberOfBatches * driverPoolConfig.singleBatchProcessTime.getSeconds) + singleBatchProcessingTempDelay
+      when (diffUTCTime now (addUTCTime (- lastScheduledTime) searchTry.startTime) < buffer) $ triggerPoolNow driverPoolConfig searchTry goHomeCfg searchReq.driverIdForSearch singleBatchProcessingTempDelay
+      scheduleJobIfPossible remainingTime driverPoolConfig.scheduleTryTimes searchTry now
+    else do
+      mbScheduleTime <- fetchAndUpdateScheduleTime driverPoolConfig searchReq now
+      case mbScheduleTime of
+        Just scheduleTime -> scheduleBatching searchTry.id scheduleTime
+        Nothing -> cancelBooking searchTry
+  where
+    cancelBooking searchTry = do
+      booking <- QRB.findByQuoteId searchTry.estimateId >>= fromMaybeM (BookingDoesNotExist searchTry.estimateId)
+      QST.updateStatus DST.CANCELLED searchTry.id
+      SBooking.cancelBooking booking Nothing merchant
+
+    getPoolTriggerScheduledKey :: CacheFlow m r => Id DSR.SearchRequest -> Int -> m (Maybe Bool)
+    getPoolTriggerScheduledKey searchReqId nextSchedule = Hedis.get (mkPoolTriggerScheduledKey searchReqId nextSchedule)
+
+    mkPoolTriggerScheduledKey :: Id DSR.SearchRequest -> Int -> Text
+    mkPoolTriggerScheduledKey searchReqId nextSchedule = "DriverPoolTrigger:SearchReqId:" <> searchReqId.getId <> ":ScheduleBeforeSeconds:" <> show nextSchedule
+
+    setPoolTriggerScheduledKey :: CacheFlow m r => Id DSR.SearchRequest -> Int -> m ()
+    setPoolTriggerScheduledKey searchReqId nextSchedule = Hedis.setExp (mkPoolTriggerScheduledKey searchReqId nextSchedule) True 1800
```

Comment:
The number 1800 occurs three times in the code. If it's the same value, it might be better to write it in a separate constant.